core/node/libp2p.go

```diff
@@ -0,0 +1,578 @@
+		You will not be able to connect to any nodes configured to use encrypted connections`)
+	return opts
+}
+
+type P2PHostIn struct {
+	fx.In
+
+	Repo          repo.Repo
+	Validator     record.Validator
+	HostOption    HostOption
+	RoutingOption RoutingOption
+	ID            peer.ID
+	Peerstore     peerstore.Peerstore
+
+	Opts [][]libp2p.Option `group:"libp2p"`
+}
+
+type BaseRouting routing.IpfsRouting
+type P2PHostOut struct {
+	fx.Out
+
+	Host    host.Host
+	Routing BaseRouting
+}
+
+func P2PHost(mctx MetricsCtx, lc fx.Lifecycle, params P2PHostIn) (out P2PHostOut, err error) {
+	opts := []libp2p.Option{libp2p.NoListenAddrs}
+	for _, o := range params.Opts {
+		opts = append(opts, o...)
+	}
+
+	ctx := lifecycleCtx(mctx, lc)
+
+	opts = append(opts, libp2p.Routing(func(h host.Host) (routing.PeerRouting, error) {
+		r, err := params.RoutingOption(ctx, h, params.Repo.Datastore(), params.Validator)
+		out.Routing = r
+		return r, err
+	}))
+
+	out.Host, err = params.HostOption(ctx, params.ID, params.Peerstore, opts...)
+	if err != nil {
+		return P2PHostOut{}, err
+	}
+
+	// this code is necessary just for tests: mock network constructions
+	// ignore the libp2p constructor options that actually construct the routing!
+	if out.Routing == nil {
+		r, err := params.RoutingOption(ctx, out.Host, params.Repo.Datastore(), params.Validator)
+		if err != nil {
+			return P2PHostOut{}, err
+		}
+		out.Routing = r
+		out.Host = routedhost.Wrap(out.Host, out.Routing)
+	}
+
+	lc.Append(fx.Hook{
+		OnStop: func(ctx context.Context) error {
+			return out.Host.Close()
+		},
+	})
+
+	return out, err
+}
+
+type Router struct {
+	routing.IpfsRouting
+
+	Priority int // less = more important
+}
+
+type p2pRouterOut struct {
+	fx.Out
+
+	Router Router `group:"routers"`
+}
+
+func P2PBaseRouting(lc fx.Lifecycle, in BaseRouting) (out p2pRouterOut, dr *dht.IpfsDHT) {
+	if dht, ok := in.(*dht.IpfsDHT); ok {
+		dr = dht
+
+		lc.Append(fx.Hook{
+			OnStop: func(ctx context.Context) error {
+				return dr.Close()
+			},
+		})
+	}
+
+	return p2pRouterOut{
+		Router: Router{
+			Priority:    1000,
+			IpfsRouting: in,
+		},
+	}, dr
+}
+
+type p2pOnlineRoutingIn struct {
+	fx.In
+
+	Routers   []Router `group:"routers"`
+	Validator record.Validator
+}
+
+func P2PRouting(in p2pOnlineRoutingIn) routing.IpfsRouting {
+	routers := in.Routers
+
+	sort.SliceStable(routers, func(i, j int) bool {
+		return routers[i].Priority < routers[j].Priority
+	})
+
+	irouters := make([]routing.IpfsRouting, len(routers))
+	for i, v := range routers {
+		irouters[i] = v.IpfsRouting
+	}
+
+	return routinghelpers.Tiered{
+		Routers:   irouters,
+		Validator: in.Validator,
+	}
+}
+
+type p2pPSRoutingIn struct {
+	fx.In
+
+	BaseRouting BaseRouting
+	Repo        repo.Repo
+	Validator   record.Validator
+	Host        host.Host
+	PubSub      *pubsub.PubSub `optional:"true"`
+}
+
+func P2PPubsubRouter(mctx MetricsCtx, lc fx.Lifecycle, in p2pPSRoutingIn) (p2pRouterOut, *namesys.PubsubValueStore) {
```

Comment:
exported func P2PPubsubRouter returns unexported type node.p2pRouterOut, which can be annoying to use